### PR TITLE
Add md5 and sha1 checksum support to Downloader.

### PR DESCRIPTION
**Pull request by chrisroberts**
Allows checksum validation on downloaded files via Util::Downloader
using MD5 and/or SHA1 checksums. This also integrates checksum validation
support with the shell provisioner for downloaded remote files.

Fixes: #6323

### CODE REVIEW

File: lib/vagrant/util/downloader.rb [hunk 1]
```diff
@@ -178,6 +191,47 @@ def head
...
+      #
+      # @param source [String] Source of file
+      # @param path [String, Pathname] local file path
+      # @param checksums [Hash] User provided options
+      # @option checksums [String] :md5 Compare MD5 checksum
+      # @option checksums [String] :sha1 Compare SHA1 checksum
+      # @return [Boolean]
+      def validate_download!(source, path, checksums)
+        CHECKSUM_MAP.each do |type, klass|
+          if checksums[type]
```
Comment (sethvargo) [inline lib/vagrant/util/downloader.rb]: Should we add some kind of log message here, like:

> Validating checksum (md5) for X, expected: xxx actual: xxx

Could help in future debugging.

File: lib/vagrant/util/downloader.rb [hunk 2]
```diff
@@ -178,6 +191,47 @@ def head
...
+      end
+
+      # Generate checksum on given file
+      #
+      # @param digest_class [Class] Digest class to use for generating checksum
+      # @param path [String, Pathname] Path to file
+      # @return [String] hexdigest result
+      def checksum_file(digest_class, path)
+        digester = digest_class.new
+        File.open(path.to_s, 'rb') do |file|
```
Comment (sethvargo) [inline lib/vagrant/util/downloader.rb]: This is 1024 _bytes_, which seems really small. The [core method](https://github.com/ruby/ruby/blob/9ec0f8b7cb49c538f2bed77ec1506d59119f4bf1/ext/digest/lib/digest.rb#L49-L57) uses 16384...

Also, there is a core method for this 😄

File: templates/locales/en.yml [hunk 3]
```diff
@@ -734,6 +734,13 @@ en:
       downloader_interrupted: |-
         The download was interrupted by an external signal. It did not
         complete.
+      downloader_checksum_error: |-
+        The calculated checksum of the requested file does not match the expected
+        checksum!
```
Comment (sethvargo) [inline templates/locales/en.yml]: Could we include some information about the type of checksum in here (md5 vs sha256)?

**Comment by chrisroberts** [comment]
@sethvargo updated

**Pull request merged**